src/uct/ib/dc/dc_mlx5_ep.c

```diff
@@ -1378,23 +1378,26 @@ void uct_dc_mlx5_ep_handle_failure(uct_dc_mlx5_ep_t *ep, void *arg,
     ucs_arbiter_group_t *group;
     uint8_t pool_index;
 
+    ucs_debug("handle failure iface: %p, dci[%d]=%x, status: %s", iface,
+              dci_index, txwq->super.qp_num, ucs_status_string(ep_status));
+
     ucs_assert(!uct_dc_mlx5_iface_is_dci_rand(iface));
 
     uct_dc_mlx5_update_tx_res(iface, txwq, txqp, pi);
     uct_rc_txqp_purge_outstanding(&iface->super.super, txqp, ep_status, pi, 0);
 
     ucs_assert(ep->dci != UCT_DC_MLX5_EP_NO_DCI);
-    uct_dc_mlx5_iface_dci_put(iface, dci_index);
     uct_dc_mlx5_get_arbiter_params(iface, ep, &waitq, &group, &pool_index);
 
     /* Do not invoke pending requests on a failed endpoint */
     ucs_arbiter_group_desched(waitq, group);
-    uct_dc_mlx5_iface_progress_pending(iface, pool_index);
-
     uct_dc_mlx5_iface_set_ep_failed(iface, ep, cqe, txwq, ep_status);
+    uct_dc_mlx5_iface_dci_put(iface, dci_index);
 
     if (ep->dci == UCT_DC_MLX5_EP_NO_DCI) {
+        /* no more operations scheduled on DCI, reset it */
         uct_dc_mlx5_iface_reset_dci(iface, dci_index);
+        uct_dc_mlx5_iface_progress_pending(iface, pool_index);
```

Comment:
can this provoke reordering now when pending progression is under if?
You release resources on line 1386, but may not invoke pending processing. Thus, some send (can be another ep) can be done OOO.

Reply:
if ep!=NO_DCI it means the resources were not actually released
IMO to clarify this can return 1/0 indication from dci_put of whether smth was released or not

Reply:
no `uct_dc_mlx5_update_tx_res` can release READ FC resources and unblock another ep sitting on another DCI

Reply:
btw, as Sergey noticed we may not need `if (ep->dci == UCT_DC_MLX5_EP_NO_DCI)` branch at all here

Reply:
right, there is call uct_rc_txqp_purge_outstanding which guarantee dci put